File: src/formats/number.js
// bool
export const boolFormater = (value, param) => {
  //if (!value) return '';
  if (param) {
    switch (param.format) {
      case 'trueFalse': return value ? 'True' : 'False';
      case 'yesNo': return value ? 'Yes' : 'No';
      case 'values':
        let values = param.values; if (!values) throw new Error('param.values undefined');
        if (values.length != 2) throw new Error('param.values invalid');
        return value ? values[0] : values[1];
      default: throw new Error('param.format invalid');
    }
  }
  return value ? 'Yes' : 'No';
};
export const boolParser = (text, param, message) => {
  if (!text) return [text, true, message];
  switch (text.toLowerCase()) {
    case '1': case 'y': case 'true': case 'yes': case 'on': return [true, true, message];
    case '0': case 'n': case 'false': case 'no': case 'off': return [false, true, message];
    default: throw new Error('invalid value');
  }
};

// decimal
export const decimalFormater = (value, param) => {
  if (!value) return '';
  value = parseFloat(value);
  if (param) {
    switch (param.format) {
      case 'pattern': return value.toFixed(param.pattern);
      default: throw new Error('param.format invalid');
    }
  }
  return value.toFixed(4);
};
export const decimalParser = (text, param, message) => {
  if (!text) return [text, true, message];
  let value = parseFloat(text); if (isNaN(value)) return [text, false, message];
  if (param) { // check param
    let minValue = param.minValue; if (minValue && value < parseFloat(minValue)) return [value, false, message];
    let maxValue = param.maxValue; if (maxValue && value > parseFloat(maxValue)) return [value, false, message];
    let precision = param.precision; if (precision && value !== Math_round(value, precision)) return [value, false, message];
    let round = param.round; if (round) value = Math_round(value, round);
  }
  return [value, true, message];
};

// integer
export const integerFormater = (value, param) => {
  if (!value) return '';
  value = parseInt(value);
  if (param) {
    switch (param.format) {
      case 'comma': return format(value);
      case 'byte':
        let length = Math.floor(parseFloat((value.toString().length - 1)) / 3);
        if (length > 0) return Math_round(value / (2 << (10 * length)), 2).toString() + ' ' + '  KBMBGB'.substring(length * 2, 2);
        if (value == 1) return '1 byte';
        return value.toString() + ' bytes';
      case 'pattern': return value.toString(param.pattern);
      default: throw new Error('param.format invalid');
    }
  }
  return value.toString();
};
export const integerParser = (text, param, message) => {
  if (!text) return [text, true, message];
  let value = parseInt(text); if (isNaN(value)) return [text, false, message];
  if (param) { // check param
    let minValue = param.minValue; if (minValue && value < parseInt(minValue)) return [value, false, message];
    let maxValue = param.maxValue; if (maxValue && value > parseInt(maxValue)) return [value, false, message];
  }
  return [value, true, message];
};

// real
export const realFormater = (value, param) => {
  if (!value) return '';
  value = parseFloat(value);
  if (param) {
    switch (param.format) {
      case 'pattern': return value.toFixed(param.pattern);
      default: throw new Error('param.format invalid');
    }
  }
  return value.toFixed(4);
};
export const realParser = (text, param, message) => {
  if (!text) return [text, true, message];
  let value = parseFloat(text); if (isNaN(value)) return [text, false, message];
  if (param) { // check param
    let minValue = param.minValue; if (minValue && value < parseFloat(minValue)) return [value, false, message];
    let maxValue = param.maxValue; if (maxValue && value > parseFloat(maxValue)) return [value, false, message];
    let precision = param.precision; if (precision && value !== Math_round(value, precision)) return [value, false, message];
    let round = param.round; if (round) value = Math_round(value, round);
  }
  return [value, true, message];
};

// money
Number.prototype.formatMoney = function (c, d, t) {
  var n = this,
    c = isNaN(c = Math.abs(c)) ? 2 : c,
    d = d == undefined ? "." : d,
    t = t == undefined ? "," : t,
    s = n < 0 ? "-" : "",
    i = String(parseInt(n = Math.abs(Number(n) || 0).toFixed(c))),
    j = (j = i.length) > 3 ? j % 3 : 0;
  return s + (j ? i.substr(0, j) + t : "") + i.substr(j).replace(/(\d{3})(?=\d)/g, "$1" + t) + (c ? d + Math.abs(n - i).toFixed(c).slice(2) : "");
};
export const moneyFormater = (value, param) => {
  if (!value) return '';
  value = parseFloat(value);
  if (param) {
    switch (param.format) {
      case 'c3': return '$' + value.formatMoney(3);
      case 'c2': return '$' + value.formatMoney(2);
      case 'pattern': return '$' + value.formatMoney(param.pattern);
      default: throw new Error('param.format invalid');
    }
  }
  return '$' + value.formatMoney(2);
};
export const moneyParser = (text, param, message) => {
  if (!text) return [text, true, message];
  //if(!/^\d+$/.test(text)) return [text, false, message];
  text = text.replace(/[^0-9\.]+/g, ''); if (!text) return [text, false, message];
  let value = parseFloat(text); if (isNaN(value)) return [text, false, message];
  value = Math_round(value, 4);
  if (param) { // check param
    let minValue = param.minValue; if (minValue && value < parseFloat(minValue)) return [value, false, message];
    let maxValue = param.maxValue; if (maxValue && value > parseFloat(maxValue)) return [value, false, message];
    let precision = param.precision; if (precision && value !== Math_round(value, precision)) return [value, false, message];
    let round = param.round; if (round) value = Math_round(value, round);
  }
  return [value, true, message];
};

// percent
export const percentFormater = (value, param) => {
  if (!value) return '';
  value = parseFloat(value);
  if (param) {
    switch (param.format) {
      case 'pattern': return value.toString(param.pattern);
      default: throw new Error('param.format invalid');
    }
  }
  return value.toString('0.00') + '%';
};
export const percentParser = (text, param, message) => {
  if (!text) return [text, true, message];
  if (text && text[text.length - 1] === '%') text = text.substring(0, text.length - 1);
  let value = parseFloat(text); if (isNaN(value)) return [text, false, message];
  return [value, true, message];
};

function format(value) {
  if (isNaN(value)) return '';
  let n = value.toString().split('.');
  return n[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',') + (n.length > 1 ? '.' + n[1] : '');
}

function Math_round(value, decimals) {
  return Number(Math.round(value + 'e' + decimals) + 'e-' + decimals);
}

File: src.cs/CoreValidation/Formats/Number.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text.RegularExpressions;
using static CoreValidation.Globals;

namespace CoreValidation.Formats
{
  public static class Number
  {
    static readonly Regex _digitPatern = new Regex(@"[^0-9\.]+", RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.Singleline);

    // bool
    public static string BoolFormater(object value, IDictionary<string, object> param = null)
    {
      if (value == null || !(value is bool value_)) return NulFormat;
      if (param != null && param.TryGetValue("format", out var z) && z is string format)
        switch (format)
        {
          case "trueFalse": return value_ ? "True" : "False";
          case "yesNo": return value_ ? "Yes" : "No";
          case "values":
            if (param.TryGetValue("values", out z) && z is string[] values && values != null)
            {
              if (values.Length != 2) throw new ArgumentOutOfRangeException(nameof(param), "param.values invalid");
              return value_ ? values[0] : values[1];
            }
            throw new ArgumentOutOfRangeException(nameof(param), "param.values undefined");
          default: throw new ArgumentOutOfRangeException(nameof(param), "param.format invalid");
        }
      return value_ ? "Yes" : "No";
    }
    public static (object, bool, Func<object>) BoolParser(string text, IDictionary<string, object> param = null, Func<object> error = null)
    {
      if (string.IsNullOrEmpty(text)) return (text, true, error);
      switch (text.ToLowerInvariant())
      {
        case "1": case "y": case "true": case "yes": case "on": return (true, true, error);
        case "0": case "n": case "false": case "no": case "off": return (false, true, error);
        default: return (text, false, error);
      }
    }

    // decimal
    public static string DecimalFormater(object value, IDictionary<string, object> param = null)
    {
      if (value == null) return NulFormat;
      var value_ = Convert.ToDecimal(value);
      if (param != null && param.TryGetValue("format", out var z) && z is string format)
        switch (format)
        {
          case "comma": return value_.ToString("#,##0", CultureInfo.InvariantCulture);
          case "n2": return value_.ToString("n2", CultureInfo.InvariantCulture);
          case "n3": return value_.ToString("n3", CultureInfo.InvariantCulture);
          case "pattern": return param.TryGetValue("pattern", out z) && z is string pattern ? value_.ToString(pattern, CultureInfo.InvariantCulture) : string.Empty;
          default: throw new ArgumentOutOfRangeException(nameof(param), "param.format invalid");
        }
      return value_.ToString("n4", CultureInfo.InvariantCulture);
    }
    public static (object, bool, Func<object>) DecimalParser(string text, IDictionary<string, object> param = null, Func<object> error = null)
    {
      if (string.IsNullOrEmpty(text)) return (text, true, error);
      if (!decimal.TryParse(text, out var value)) return (text, false, error);
      if (param != null)
      { // check param
        if (param.TryGetValue("minValue", out var z) && z is decimal minValue) { if (minValue != 0 && value < minValue) return (text, false, error); }
        if (param.TryGetValue("maxValue", out z) && z is decimal maxValue) { if (maxValue != 0 && value > maxValue) return (text, false, error); }
        if (param.TryGetValue("precision", out z) && z is int precision) { if (precision != 0 && value < Math.Round(value, precision)) return (text, false, error); }
        if (param.TryGetValue("round", out z) && z is int round) { if (round != 0) value = Math.Round(value, round); }
      }
      return (value, true, error);
    }

    // integer
    public static string IntegerFormater(object value, IDictionary<string, object> param = null)
    {
      if (value == null) return NulFormat;
      var value_ = Convert.ToInt32(value);
      if (param != null && param.TryGetValue("format", out var z) && z is string format)
        switch (format)
        {
          case "comma": return value_.ToString("#,##0", CultureInfo.InvariantCulture);
          case "byte":
            var radix = (int)Math.Floor((float)(value_.ToString().Length - 1) / 3);
            if (radix > 0) return $"{Math.Round((double)(value_ / (1 << 10 * radix)), 2)} {"  KBMBGB".Substring(radix << 1, 2)}";
            if (value_ == 1) return "1 byte";
            return $"{value_} bytes";
          case "pattern": return param.TryGetValue("pattern", out z) && z is string pattern ? value_.ToString(pattern, CultureInfo.InvariantCulture) : string.Empty;
          default: throw new ArgumentOutOfRangeException(nameof(param), "param.format invalid");
        }
      return value_.ToString(CultureInfo.InvariantCulture);
    }
    public static (object, bool, Func<object>) IntegerParser(string text, IDictionary<string, object> param = null, Func<object> error = null)
    {
      if (string.IsNullOrEmpty(text)) return (text, true, error);
      if (!int.TryParse(text, out var value)) return (text, false, error);
      if (param != null)
      { // check param
        if (param.TryGetValue("minValue", out var z) && z is decimal minValue) { if (minValue != 0 && value < minValue) return (text, false, error); }
        if (param.TryGetValue("maxValue", out z) && z is decimal maxValue) { if (maxValue != 0 && value > maxValue) return (text, false, error); }
      }
      return (value, true, error);
    }

    // real
    public static string RealFormater(object value, IDictionary<string, object> param = null)
    {
      if (value == null) return NulFormat;
      var value_ = Convert.ToSingle(value);
      if (param != null && param.TryGetValue("format", out var z) && z is string format)
        switch (format)
        {
          case "comma": return value_.ToString("#,##0", CultureInfo.InvariantCulture);
          case "n2": return value_.ToString("n2", CultureInfo.InvariantCulture);
          case "n3": return value_.ToString("n3", CultureInfo.InvariantCulture);
          case "pattern": return param.TryGetValue("pattern", out z) && z is string pattern ? value_.ToString(pattern, CultureInfo.InvariantCulture) : string.Empty;
          default: throw new ArgumentOutOfRangeException(nameof(param), "param.format invalid");
        }
      return value_.ToString("n4", CultureInfo.InvariantCulture);
    }
    public static (object, bool, Func<object>) RealParser(string text, IDictionary<string, object> param = null, Func<object> error = null)
    {
      if (string.IsNullOrEmpty(text)) return (text, true, error);
      if (!float.TryParse(text, out var value)) return (text, false, error);
      if (param != null)
      { // check param
        if (param.TryGetValue("minValue", out var z) && z is float minValue) { if (minValue != 0 && value < minValue) return (text, false, error); }
        if (param.TryGetValue("maxValue", out z) && z is float maxValue) { if (maxValue != 0 && value > maxValue) return (text, false, error); }
        if (param.TryGetValue("precision", out z) && z is int precision) { if (precision != 0 && value < Math.Round(value, precision)) return (text, false, error); }
        if (param.TryGetValue("round", out z) && z is int round) { if (round != 0) value = (float)Math.Round(value, round); }
      }
      return (value, true, error);
    }

    // money
    public static string MoneyFormater(object value, IDictionary<string, object> param = null)
    {
      if (value == null) return NulFormat;
      var value_ = Convert.ToDecimal(value);
      if (param != null && param.TryGetValue("format", out var z) && z is string format)
        switch (format)
        {
          case "c2": return value_.ToString("c2", CultureInfo.InvariantCulture);
          case "c3": return value_.ToString("c3", CultureInfo.InvariantCulture);
          case "pattern": return param.TryGetValue("pattern", out z) && z is string pattern ? value_.ToString(pattern, CultureInfo.InvariantCulture) : string.Empty;
          default: throw new ArgumentOutOfRangeException(nameof(param), "param.format invalid");
        }
      return value_.ToString("c2", CultureInfo.InvariantCulture);
    }
    public static (object, bool, Func<object>) MoneyParser(string text, IDictionary<string, object> param = null, Func<object> error = null)
    {
      if (string.IsNullOrEmpty(text)) return (text, true, error);
      text = _digitPatern.Replace(text, string.Empty); if (string.IsNullOrEmpty(text)) return (text, true, error);
      if (!decimal.TryParse(text, out var value)) return (text, false, error);
      value = Math.Round(value, 4);
      if (param != null)
      { // check param
        if (param.TryGetValue("minValue", out var z) && z is decimal minValue) { if (minValue != 0 && value < minValue) return (text, false, error); }
        if (param.TryGetValue("maxValue", out z) && z is decimal maxValue) { if (maxValue != 0 && value > maxValue) return (text, false, error); }
        if (param.TryGetValue("precision", out z) && z is int precision) { if (precision != 0 && value < Math.Round(value, precision)) return (text, false, error); }
        if (param.TryGetValue("round", out z) && z is int round) { if (round != 0) value = Math.Round(value, round); }
      }
      return (value, true, error);
    }

    // percent
    public static string PercentFormater(object value, IDictionary<string, object> param = null)
    {
      if (value == null) return NulFormat;
      var value_ = Convert.ToSingle(value);
      if (param != null && param.TryGetValue("format", out var z) && z is string format)
        switch (format)
        {
          case "p2": return value_.ToString("p2", CultureInfo.InvariantCulture);
          case "p3": return value_.ToString("p3", CultureInfo.InvariantCulture);
          case "pattern": return param.TryGetValue("pattern", out z) && z is string pattern ? value_.ToString(pattern, CultureInfo.InvariantCulture) : string.Empty;
          default: throw new ArgumentOutOfRangeException(nameof(param), "param.format invalid");
        }
      return value_.ToString("p2", CultureInfo.InvariantCulture);
    }
    public static (object, bool, Func<object>) PercentParser(string text, IDictionary<string, object> param = null, Func<object> error = null)
    {
      if (string.IsNullOrEmpty(text)) return (text, true, error);
      if (text[text.Length - 1] == '%') text = text.Substring(0, text.Length - 1);
      if (!float.TryParse(text, out var value)) return (text, false, error);
      return (value / 100, true, error);
    }
  }
}
